File: Programmers/C++/외벽 점검_2.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#define FOR(i,n) for(int i=0;i<(n);++i)
#define FORS(i,s,n) for(int i=(s);i<(n);++i)
#define INF 987654321

using namespace std;

int res = INF;

int solution(int n, vector<int> weak, vector<int> dist) {
    int w = weak.size();
    FOR(i,w) weak.push_back(n+weak[i]);
    sort(dist.begin(), dist.end());
    
    do {
        FOR(i, w) {
            int st = weak[i];
            int ed = weak[i+w-1];
            
            FOR(j, dist.size()) {
                st += dist[j];
                if(st>=ed) {
                    res = min(res, j+1);
                    break;
                }
                int next = j+1;
                while(weak[next]<=st) ++next;
                st = weak[next];
            }
        }
    } while(next_permutation(dist.begin(), dist.end()));
        
    return res==INF? -1 : res;
}

File: Programmers/Swift/외벽 점검_2.swift
import Foundation

let INF = 987654321
var res = 0
var w = [Int]()
var wSz = 0
var d = [Int]()
var visited = [Bool](repeating:false, count: 8)

func perm(_ idx: inout [Int]) {
    if idx.count == d.count {
        for i in 0..<wSz {
            var st = w[i]
            let ed = w[i+wSz-1]
            for j in 0..<d.count {
                st += d[idx[j]]
                if st>=ed {
                    if res > j+1 {res=j+1}
                    break
                }
                var next = wSz-1
                var l = i+1
                var r = wSz-1
                while l<=r {
                    let mid = (l+r)/2
                    if w[mid] > st {
                        if next > mid {next=mid}
                        r = mid-1
                    }
                    else {l = mid+1}
                }
                st = w[next]
            }
        }
        return
    }
    for i in 0..<d.count {
        if visited[i]==true {continue}
        visited[i] = true
        idx.append(i)
        perm(&idx)
        visited[i] = false
        idx.removeLast()
    }
}

func solution(_ n:Int, _ weak:[Int], _ dist:[Int]) -> Int {
    
    res = INF
    w = weak
    wSz = weak.count
    
    for i in 0..<wSz {
        w.append(n+w[i])
    }
    
    d = dist
    var idx = [Int]()
    perm(&idx)
    return res==INF ? -1 : res
}
